### PR TITLE
Immigration

### PR DESCRIPTION
**Pull request by jpawlicki**
• Implements a new mechanic for population motion between regions.
 • Fixes a potential infinite loop in pirate spawning.

### CODE REVIEW

File: backend/src/main/java/com/empire/World.java [hunk 1]
```diff
@@ -2064,6 +2047,66 @@ void eatFood() {
 		}
 
 		void growPopulations() {
+			// Emigration
+			class Emigration {
+				final Region source;
+				final Region sink;
```
Comment (Alex-Pawlicki-Bose) [inline backend/src/main/java/com/empire/World.java]: Mostly a semantic argument being made here...source/sink usually means the creation (seemingly out of thin air) and subsequent consumption of some object/quantity.  I think source/destination might be a more accurate description here.  Doesn't really matter though.

File: backend/src/main/java/com/empire/World.java [hunk 2]
```diff
@@ -2064,6 +2047,66 @@ void eatFood() {
...
+				}
+			}
+			List<Emigration> emigrations = new ArrayList<>();
+			for (Region r : regions) {
+				double eligibleToEmigrate = r.population * r.unrestPopular * Constants.emigrationFactor;
+				if (starvingRegions.contains(r)) eligibleToEmigrate *= (1 + Constants.emigrationStarvationMod);
+				eligibleToEmigrate = Math.min(eligibleToEmigrate, r.population - 1); // Emigrating the last person causes a divide by zero error.
+				List<Region> destinations = new ArrayList<Region>();
+				for (Region n : r.getNeighbors(World.this)) {
+					if (n.type == Region.Type.LAND) destinations.add(n);
```
Comment (Alex-Pawlicki-Bose) [inline backend/src/main/java/com/empire/World.java]: Could use region.isLand() methods here and a few lines down as well.

File: backend/src/main/java/com/empire/World.java [hunk 3]
```diff
@@ -2064,6 +2047,66 @@ void eatFood() {
...
+					else for (Region nn : n.getNeighbors(World.this)) {
+						if (nn != r && n.type == Region.Type.LAND) destinations.add(nn);
```
Comment (Alex-Pawlicki-Bose) [inline backend/src/main/java/com/empire/World.java]: This seems to imply that destinations can be a neighbor of a neighbor to the region r being considered, hence not necessarily adjacent.  Just checking that this is intentional.

File: backend/src/main/java/com/empire/World.java [hunk 4]
```diff
@@ -2406,7 +2449,6 @@ void advanceDate() {
 
 	private void addPopulation(Region r, double amount) {
 		r.population += amount;
-		int rid = regions.indexOf(r);
```
Comment (Alex-Pawlicki-Bose) [inline backend/src/main/java/com/empire/World.java]: This comment is not really related to the change made, but this method really looks like it should belong to the Region class.

File: backend/src/main/java/com/empire/World.java [hunk 5]
```diff
@@ -2064,6 +2047,66 @@ void eatFood() {
...
+				destinations.removeIf(d -> !d.getKingdom().equals(r.getKingdom()) && (patrolledRegions.contains(r) || getNation(d.getKingdom()).getRelationship(r.getKingdom()).refugees == Relationship.Refugees.REFUSE));
+				if (destinations.isEmpty() || eligibleToEmigrate < 1) continue;
+				double totalWeight = 0;
+				for (Region d : destinations) totalWeight += d.calcImmigrationWeight();
+				for (Region d : destinations) {
+					emigrations.add(new Emigration(r, d, d.calcImmigrationWeight() / totalWeight * eligibleToEmigrate));
+				}
+			}
+			Map<Region, Double> netMotion = new HashMap<>();
+			for (Emigration e : emigrations) {
```
Comment (Alex-Pawlicki-Bose) [inline backend/src/main/java/com/empire/World.java]: I would consider moving most of the inner part of this loop to the Emigration class as a sort of apply() method for doing all the population updating, and keep the netMotion-related calculation outside as is.

File: backend/src/main/java/com/empire/World.java [hunk 6]
```diff
@@ -2064,6 +2047,66 @@ void eatFood() {
...
+				netMotion.merge(e.source, -e.population, Double::sum);
+				double unhappyCitizens = e.source.population * e.source.unrestPopular;
+				e.source.population -= e.population;
+				e.source.unrestPopular = (unhappyCitizens - e.population) / e.source.population;
+				netMotion.merge(e.sink, e.population, Double::sum);
+				unhappyCitizens = e.sink.population * e.sink.unrestPopular;
+				e.sink.population += e.population;
+				e.sink.unrestPopular = (unhappyCitizens + e.population) / e.sink.population;
+			}
+			for (String k : kingdoms.keySet()) {
```
Comment (Alex-Pawlicki-Bose) [inline backend/src/main/java/com/empire/World.java]: From the player's point of view, this could be really frustrating not knowing more of the particulars around where population is coming from and being lost if you are only notified of the net changes.  Without slicing it too finely numbers-wise, just having a hint of where positive/negative flows are coming from would probably suffice.

File: backend/src/main/java/com/empire/World.java (hunk 1 above)
Comment (jpawlicki) [reply backend/src/main/java/com/empire/World.java]: Done

File: backend/src/main/java/com/empire/World.java (hunk 2 above)
Comment (jpawlicki) [reply backend/src/main/java/com/empire/World.java]: Done.

File: backend/src/main/java/com/empire/World.java (hunk 3 above)
Comment (jpawlicki) [reply backend/src/main/java/com/empire/World.java]: Yes, but only if r is a water region.

File: backend/src/main/java/com/empire/World.java (hunk 4 above)
Comment (jpawlicki) [reply backend/src/main/java/com/empire/World.java]: Yeah.

File: backend/src/main/java/com/empire/World.java (hunk 5 above)
Comment (jpawlicki) [reply backend/src/main/java/com/empire/World.java]: Done.

File: backend/src/main/java/com/empire/World.java (hunk 6 above)
Comment (jpawlicki) [reply backend/src/main/java/com/empire/World.java]: Hm, you do get the details of which regions go up and down, I think the next step would be to tell you each of the emigrations - not sure if that's the too much detail you're talking about. Do you have an example of what you have in mind?

File: backend/src/main/java/com/empire/World.java (hunk 3 above)
Comment (Alex-Pawlicki-Bose) [reply backend/src/main/java/com/empire/World.java]: Ah you're right, I didn't process that correctly.

File: backend/src/main/java/com/empire/World.java (hunk 6 above)
Comment (Alex-Pawlicki-Bose) [reply backend/src/main/java/com/empire/World.java]: I was thinking additional info along the lines of "You gained pop. from regions A, B, C. You lost pop. to regions X, Y, Z"

Comment (jpawlicki) [reply backend/src/main/java/com/empire/World.java]: Added more detail.
